File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/MessageTree.java
package io.euphoria.xkcd.app.impl.ui;

import android.os.Parcel;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * @author N00bySumairu
 */

public class MessageTree implements Comparable<MessageTree> {

    public static long idStringToLong(String id) {
        if (id.equals(CURSOR_ID)) return -1;
        return Long.parseLong(id, 36);
    }

    public static String idLongToString(long id) {
        if (id == -1) return CURSOR_ID;
        return Long.toString(id, 36);
    }

    private static String formatID(String id) {
        return (id == null) ? "-" : (id.equals(CURSOR_ID)) ? "~" : id;
    }

    public static final String CURSOR_ID = "\uFFFF";

    protected static final byte PF_IS_A_THING  = 0x01;
    protected static final byte PF_HAS_CONTENT = 0x02;
    protected static final byte PF_HAS_REPLIES = 0x04;
    protected static final byte PF_TRUNCATED   = 0x08;
    protected static final byte PF_COLLAPSED   = 0x10;

    private final List<MessageTree> replies = new ArrayList<>();
    private String id;
    private String parent;
    private UIMessage message;
    private long longID;
    private int indent = 0;
    private boolean collapsed = false;

    public MessageTree(UIMessage m) {
        message = m;
        if (m == null) {
            id = CURSOR_ID;
            parent = null;
        } else {
            id = m.getID();
            parent = m.getParent();
        }
        longID = idStringToLong(id);
    }

    protected MessageTree(Parcel in, byte flags, String parent) {
        if ((flags & PF_IS_A_THING) == 0) {
            throw new IllegalArgumentException("Invalid encoded MessageTree flags");
        }
        String id = in.readString();
        if ((flags & PF_HAS_CONTENT) != 0) {
            message = new UIMessage(in, id, parent, ((flags & PF_TRUNCATED) != 0));
        }
        this.id = id;
        this.parent = parent;
        this.longID = idStringToLong(id);
        this.collapsed = ((flags & PF_COLLAPSED) != 0);
        if ((flags & PF_HAS_REPLIES) != 0) {
            while (true) {
                byte nextFlags = in.readByte();
                if (nextFlags == 0) break;
                addReply(new MessageTree(in, nextFlags, id));
            }
        }
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof MessageTree && compareTo((MessageTree) o) == 0;
    }

    @Override
    public String toString() {
        return String.format("%s@%x[%s/%s]", getClass().getSimpleName(), hashCode(), formatID(parent), formatID(id));
    }

    @Override
    public int compareTo(@NonNull MessageTree o) {
        return id.compareTo(o.id);
    }

    /** The ID of this MessageTree (CURSOR_ID for the input bar). */
    public String getID() {
        return id;
    }

    /** The parent node of this MessageTree. */
    public String getParent() {
        return parent;
    }

    /**
     * A long containing the numerical value of this message's ID.
     * CURSOR_ID is mapped to the special value -1.
     */
    public long getLongID() {
        return longID;
    }

    /**
     * Set this MessageTree's parent ID.
     * Only permissible for a MessageTree representing the input bar.
     */
    public void setParent(String parent) {
        if (message != null) throw new IllegalStateException("Attempting to change parent of message");
        this.parent = parent;
    }

    /** Sorted list of the immediate replies of this tree. */
    public List<MessageTree> getReplies() {
        return Collections.unmodifiableList(replies);
    }

    /** The message wrapped by this; null for the input bar. */
    public UIMessage getMessage() {
        return message;
    }

    /** Change the message wrapped by this MessageTree. */
    public void setMessage(@NonNull UIMessage m) {
        assert m.getID().equals(id) : "Updating MessageTree with unrelated message";
        message = m;
        id = m.getID();
        parent = m.getParent();
        longID = idStringToLong(id);
    }

    /** The indentation level of this. */
    public int getIndent() {
        return indent;
    }

    protected void updateIndent(int i) {
        indent = i++;
        for (MessageTree t : replies) t.updateIndent(i);
    }

    /** Whether the <em>replies</em> of this are invisible. */
    public boolean isCollapsed() {
        return collapsed;
    }

    /** Set the {@link #isCollapsed()} flag. */
    public void setCollapsed(boolean collapsed) {
        this.collapsed = collapsed;
    }

    /**
     * Add a MessageTree to the replies list.
     * Returns the index into the replies list at which t now resides.
     */
    public int addReply(@NonNull MessageTree t) {
        if (message == null) throw new IllegalStateException("Input bar cannot have replies");
        t.updateIndent(indent + 1);
        return UIUtils.insertSorted(replies, t);
    }

    /** Add every MessageTree in the list as a reply. */
    public void addReplies(@NonNull Collection<MessageTree> list) {
        // TODO replace with bulk insert, sort, and deduplication?
        for (MessageTree t : list) addReply(t);
    }

    /** Remove a MessageTree from the replies list. */
    public void removeReply(@NonNull MessageTree t) {
        UIUtils.removeSorted(replies, t);
    }

    /**
     * Return the amount of visible replies to this MessageTree.
     * A MessageTree is visible iff it has no parent or its parent is visible and not collapsed.
     * The message this method is invoked upon is assumed not to have a parent.
     */
    public int countVisibleReplies() {
        // TODO cache this?
        if (collapsed) return 0;
        int ret = 0;
        for (MessageTree mt : replies) {
            ret += 1 + mt.countVisibleReplies();
        }
        return ret;
    }

    /**
     * Count replies for user display
     * If override is true, this and only this MessageTree is assumed to be visible regardless of its actual state.
     * The input bar does not count; otherwise equivalent to countVisibleReplies().
     */
    public int countVisibleUserReplies(boolean override) {
        if (!override && collapsed) return 0;
        int ret = 0;
        for (MessageTree mt : replies) {
            if (mt.getID().equals(CURSOR_ID)) continue;
            ret += 1 + mt.countVisibleUserReplies(false);
        }
        return ret;
    }

    /** Return a list of all the visible replies to this MessageTree. */
    public List<MessageTree> traverseVisibleReplies(boolean includeThis) {
        List<MessageTree> ret = new ArrayList<>();
        if (includeThis) ret.add(this);
        traverseVisibleRepliesInner(ret);
        return ret;
    }

    private void traverseVisibleRepliesInner(List<MessageTree> drain) {
        if (isCollapsed()) return;
        for (MessageTree t : replies) {
            drain.add(t);
            t.traverseVisibleRepliesInner(drain);
        }
    }

    /** Serialization logic. */
    protected void writeToParcel(Parcel out) {
        byte flags = PF_IS_A_THING;
        if (message != null) {
            flags |= PF_HAS_CONTENT;
            if (message.isTruncated()) flags |= PF_TRUNCATED;
        }
        if (replies.size() != 0) {
            flags |= PF_HAS_REPLIES;
        }
        if (isCollapsed()) {
            flags |= PF_COLLAPSED;
        }
        out.writeByte(flags);
        out.writeString(id);
        if ((flags & PF_HAS_CONTENT) != 0) {
            message.writeToParcel(out);
        }
        if ((flags & PF_HAS_REPLIES) != 0) {
            for (MessageTree mt : replies) mt.writeToParcel(out);
            out.writeByte((byte) 0);
        }
    }

}

File: app/src/main/java/io/euphoria/xkcd/app/impl/ui/data/MessageTree.kt
package io.euphoria.xkcd.app.impl.ui.data

import android.os.Parcel
import io.euphoria.xkcd.app.hasFlags
import io.euphoria.xkcd.app.impl.ui.UIUtils
import io.euphoria.xkcd.app.withFlags
import java.util.*

/**
 * @author N00bySumairu
 */
class MessageTree : Comparable<MessageTree> {
    private val _replies: MutableList<MessageTree> = ArrayList()
    val replies: List<MessageTree>
        /** Sorted list of the immediate replies of this tree.  */
        get() = Collections.unmodifiableList(_replies)
    /** The ID of this MessageTree (CURSOR_ID for the input bar).  */
    lateinit var id: String
        private set

    private var _parent: String? = null
    /** The id of the parent node of this MessageTree.  */
    var parent: String?
        get() = _parent
        /**
         * Set this MessageTree's parent ID.
         * Only permissible for a MessageTree representing the input bar.
         */
        set(parent) {
            check(message == null) { "Attempting to change parent of message" }
            _parent = parent
        }
    /** The message wrapped by this; null for the input bar.  */
    var message: UIMessage? = null
        private set

    /**
     * A long containing the numerical value of this message's ID.
     * CURSOR_ID is mapped to the special value -1.
     */
    var longID: Long
        private set
    /** The indentation level of this.  */
    var indent = 0
        private set
    /** Whether the *replies* of this are invisible.  */
    /** Set the [.isCollapsed] flag.  */
    var isCollapsed = false

    constructor(m: UIMessage?) {
        message = m
        if (m == null) {
            id = CURSOR_ID
            _parent = null
        } else {
            id = m.id
            _parent = m.parent
        }
        longID = idStringToLong(id)
    }

    constructor(`in`: Parcel, flags: Byte, parent: String?) {
        require(flags.hasFlags(PF_IS_A_THING)) { "Invalid encoded MessageTree flags" }
        val id = `in`.readString()!!
        if (flags.hasFlags(PF_HAS_CONTENT)) {
            message =
                UIMessage(`in`, id, parent, flags.hasFlags(PF_TRUNCATED))
        }
        this.id = id
        this._parent = parent
        longID = idStringToLong(id)
        isCollapsed = flags.hasFlags(PF_COLLAPSED)
        if (flags.hasFlags(PF_HAS_REPLIES)) {
            while (true) {
                val nextFlags = `in`.readByte()
                if (nextFlags.toInt() == 0) break
                addReply(MessageTree(`in`, nextFlags, id))
            }
        }
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return other is MessageTree && compareTo(other) == 0
    }

    override fun toString(): String {
        return String.format(
            "%s@%x[%s/%s]",
            javaClass.simpleName,
            hashCode(),
            formatID(_parent),
            formatID(id)
        )
    }

    override fun compareTo(other: MessageTree): Int {
        return id.compareTo(other.id)
    }

    /** Change the message wrapped by this MessageTree.  */
    fun setMessage(m: UIMessage) {
        assert(m.id == id) { "Updating MessageTree with unrelated message" }
        message = m
        id = m.id
        _parent = m.parent
        longID = idStringToLong(id)
    }

    fun updateIndent(i: Int) {
        indent = i
        for (t in _replies) t.updateIndent(i + 1)
    }

    /**
     * Add a MessageTree to the replies list.
     * Returns the index into the replies list at which t now resides.
     */
    fun addReply(t: MessageTree): Int {
        checkNotNull(message) { "Input bar cannot have replies" }
        t.updateIndent(indent + 1)
        return UIUtils.insertSorted(_replies, t)
    }

    /** Add every MessageTree in the list as a reply.  */
    fun addReplies(list: Collection<MessageTree>) {
        // TODO replace with bulk insert, sort, and deduplication?
        for (t in list) addReply(t)
    }

    /** Remove a MessageTree from the replies list.  */
    fun removeReply(t: MessageTree) {
        UIUtils.removeSorted(_replies, t)
    }

    /**
     * Return the amount of visible replies to this MessageTree.
     * A MessageTree is visible iff it has no parent or its parent is visible and not collapsed.
     * The message this method is invoked upon is assumed not to have a parent.
     */
    fun countVisibleReplies(): Int {
        // TODO cache this?
        if (isCollapsed) return 0
        var ret = 0
        for (mt in _replies) {
            ret += 1 + mt.countVisibleReplies()
        }
        return ret
    }

    /**
     * Count replies for user display
     * If override is true, this and only this MessageTree is assumed to be visible regardless of its actual state.
     * The input bar does not count; otherwise equivalent to countVisibleReplies().
     */
    fun countVisibleUserReplies(override: Boolean): Int {
        if (!override && isCollapsed) return 0
        var ret = 0
        for (mt in _replies) {
            if (mt.id == CURSOR_ID) continue
            ret += 1 + mt.countVisibleUserReplies(false)
        }
        return ret
    }

    /** Return a list of all the visible replies to this MessageTree.  */
    fun traverseVisibleReplies(includeThis: Boolean): List<MessageTree> {
        val ret: MutableList<MessageTree> = ArrayList()
        if (includeThis) ret.add(this)
        traverseVisibleRepliesInner(ret)
        return ret
    }

    private fun traverseVisibleRepliesInner(drain: MutableList<MessageTree>) {
        if (isCollapsed) return
        for (t in _replies) {
            drain.add(t)
            t.traverseVisibleRepliesInner(drain)
        }
    }

    /** Serialization logic.  */
    fun writeToParcel(out: Parcel) {
        var flags = PF_IS_A_THING
        if (message != null) {
            flags = flags.withFlags(PF_HAS_CONTENT)
            if (message!!.isTruncated) flags = flags.withFlags(PF_TRUNCATED)
        }
        if (_replies.size != 0) {
            flags = flags.withFlags(PF_HAS_REPLIES)
        }
        if (isCollapsed) {
            flags = flags.withFlags(PF_COLLAPSED)
        }
        out.writeByte(flags)
        out.writeString(id)
        if (flags.hasFlags(PF_HAS_CONTENT)) {
            message!!.writeToParcel(out)
        }
        if (flags.hasFlags(PF_HAS_REPLIES)) {
            for (mt in _replies) mt.writeToParcel(out)
            out.writeByte(0.toByte())
        }
    }

    companion object {
        fun idStringToLong(id: String): Long {
            return if (id == CURSOR_ID) -1 else id.toLong(36)
        }

        fun idLongToString(id: Long): String {
            return if (id == -1L) CURSOR_ID else id.toString(36)
        }

        private fun formatID(id: String?): String {
            return if (id == null) "-" else if (id == CURSOR_ID) "~" else id
        }

        const val CURSOR_ID = "\uFFFF"
        const val PF_IS_A_THING: Byte = 0x01
        protected const val PF_HAS_CONTENT: Byte = 0x02
        protected const val PF_HAS_REPLIES: Byte = 0x04
        protected const val PF_TRUNCATED: Byte = 0x08
        protected const val PF_COLLAPSED: Byte = 0x10
    }
}
